Remotion/Web/Development.WebTesting/WebTestSetUpFixtureHelper.cs

```diff
@@ -107,9 +120,61 @@ private void HostWebApplication ()
       _hostingStrategy.DeployAndStartWebApplication();
     }
 
+    private void VerifyWebApplicationStarted (string webApplicationRoot, TimeSpan applicationPingTimeout)
+    {
+      var stopwatch = Stopwatch.StartNew();
+      var webRequest = (HttpWebRequest) HttpWebRequest.Create (webApplicationRoot);
+      webRequest.Method = WebRequestMethods.Http.Head;
+      webRequest.AllowAutoRedirect = true;
+      HttpStatusCode statusCode = default;
+      Assertion.DebugAssert (statusCode != HttpStatusCode.OK);
+
+      while (statusCode != HttpStatusCode.OK)
+      {
+        try
+        {
+          var responseTask = webRequest.GetResponseAsync();
+          if (responseTask.Wait (applicationPingTimeout - stopwatch.Elapsed))
+          {
+            using (var response = (HttpWebResponse) responseTask.Result)
+            {
+              statusCode = HttpStatusCode.NotFound;
+            }            
+          }
+          else
+          {
+            responseTask.Dispose();
+            
+            CheckTimeout (webApplicationRoot, applicationPingTimeout, stopwatch, $"Failed because the WebRequest timed out."); 
+          }
+        }
+        catch (WebException ex)
+        {
+          CheckTimeout (webApplicationRoot, applicationPingTimeout, stopwatch, $"Failed with WebException '{ex.Message}'"); 
+        }
+
+        CheckTimeout (webApplicationRoot, applicationPingTimeout, stopwatch, $"Failed with HttpStatusCode '{statusCode}'");
+
+        Thread.Sleep (TimeSpan.FromMilliseconds (500));
+      }
+    }
+
     private void UnhostWebApplication ()
     {
       _hostingStrategy.StopAndUndeployWebApplication();
     }
+
+    // ReSharper disable once ParameterOnlyUsedForPreconditionCheck.Local
+    private void CheckTimeout (string webApplicationRoot, TimeSpan applicationPingTimeout, Stopwatch stopwatch, string failureReason)
+    {
+      if (stopwatch.ElapsedMilliseconds > applicationPingTimeout.TotalMilliseconds)
+      {
+        throw new InvalidOperationException (

```

Comment:
Should this be a WebException?

Reply:
What about a `TimeoutException`?

Reply:
Depends. If we want to represent a webrequest with a timeout, then webexception with a timeout result coude would be more correct. What does HttpWebRequest say about timeouts?

Reply:
> What does HttpWebRequest say about timeouts?

Concerning what aspect?

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.net.httpwebrequest.getresponse?view=netframework-4.8

> WebException
> 
> Abort() was previously called.
> 
> -or-
> 
> The time-out period for the request expired.
> 
> -or-
> 
> An error occurred while processing the request.
>

Reply:
I agree that we are polling the endpoint using a webrequest, and there is a timeout, but the `WebException(..., WebExceptionStatus.Timeout)` does represent a **single request** that ran into a timeout, while there is the possibility that multiple requests have failed & the **polling** ran into a timeout -> would this still be a `WebException(..., WebExceptionStatus.Timeout)`?

Reply:
It's a matter of abstraction. If we want to hide the details and just say that the "Server doesn't answer within the timeout" it's a WebException. If we want to say that "something ran in to a timeout", we could also do a TimeoutException instead.

Reply:
Sold. (a4e38d5fa1e107212dc90ce9f40a8b989809120e)